NettyMessagingServiceTest.testSendAndReceiveWithFixedTimeout

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? Seen once so far
- Does it block your work? yes
- Do we suspect that it is a real failure? unknown

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Error Message

io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use

Stacktrace

java.util.concurrent.CompletionException: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
	at java.base/java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:777)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$bind$32(NettyMessagingService.java:699)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:500)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:474)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:413)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:538)
	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:531)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:111)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetFailure(AbstractChannel.java:991)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:553)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1345)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:503)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:488)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:984)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:247)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:355)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:416)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:331)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use

</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>
22:15:31.072 [] INFO  io.zeebe.test - Test started: testSendAndReceiveWithFixedTimeout(io.atomix.cluster.messaging.impl.NettyMessagingServiceTest)
22:15:32.076 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - TCP server listening for connections on 0.0.0.0:37733
22:15:32.159 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - Started
22:15:32.174 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - TCP server listening for connections on 0.0.0.0:40295
22:15:32.175 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - Started
22:15:32.274 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - TCP server listening for connections on 0.0.0.0:45217
22:15:32.276 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - Started
22:15:32.359 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - TCP server listening for connections on 0.0.0.0:37025
22:15:32.361 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - Started
22:15:32.455 [] WARN  io.atomix.cluster.messaging.impl.NettyMessagingService - Failed to bind TCP server to port 0.0.0.0:32797 due to io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
22:15:34.960 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - Stopped
22:15:37.173 [] INFO  io.atomix.cluster.messaging.impl.NettyMessagingService - Stopped
</pre>
</details>

--- a/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyBroadcastServiceTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyBroadcastServiceTest.java
@@ -20,8 +20,7 @@ import static org.slf4j.LoggerFactory.getLogger;
 
 import io.atomix.cluster.messaging.ManagedBroadcastService;
 import io.atomix.utils.net.Address;
-import java.io.IOException;
-import java.net.ServerSocket;
+import io.zeebe.test.util.socket.SocketUtil;
 import net.jodah.concurrentunit.ConcurrentTestCase;
 import org.junit.After;
 import org.junit.Before;
@@ -55,9 +54,9 @@ public class NettyBroadcastServiceTest extends ConcurrentTestCase {
 
   @Before
   public void setUp() throws Exception {
-    localAddress1 = Address.from("127.0.0.1", findAvailablePort(5001));
-    localAddress2 = Address.from("127.0.0.1", findAvailablePort(5002));
-    groupAddress = Address.from("230.0.0.1", findAvailablePort(1234));
+    localAddress1 = Address.from("127.0.0.1", SocketUtil.getNextAddress().getPort());
+    localAddress2 = Address.from("127.0.0.1", SocketUtil.getNextAddress().getPort());
+    groupAddress = Address.from("230.0.0.1", SocketUtil.getNextAddress().getPort());
 
     netty1 =
         (ManagedBroadcastService)
@@ -94,18 +93,6 @@ public class NettyBroadcastServiceTest extends ConcurrentTestCase {
       } catch (final Exception e) {
         LOGGER.warn("Failed stopping netty2", e);
       }
-    }
-  }
-
-  private static int findAvailablePort(final int defaultPort) {
-    try {
-      final ServerSocket socket = new ServerSocket(0);
-      socket.setReuseAddress(true);
-      final int port = socket.getLocalPort();
-      socket.close();
-      return port;
-    } catch (final IOException ex) {
-      return defaultPort;
     }
   }
 }

--- a/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java
@@ -30,9 +30,8 @@ import com.google.common.util.concurrent.Uninterruptibles;
 import io.atomix.cluster.messaging.ManagedMessagingService;
 import io.atomix.cluster.messaging.MessagingConfig;
 import io.atomix.utils.net.Address;
-import java.io.IOException;
+import io.zeebe.test.util.socket.SocketUtil;
 import java.net.ConnectException;
-import java.net.ServerSocket;
 import java.time.Duration;
 import java.util.Arrays;
 import java.util.UUID;
@@ -73,38 +72,38 @@ public class NettyMessagingServiceTest {
 
   @Before
   public void setUp() throws Exception {
-    address1 = Address.from(findAvailablePort(5001));
+    address1 = Address.from(SocketUtil.getNextAddress().getPort());
     netty1 =
         (ManagedMessagingService)
             new NettyMessagingService("test", address1, new MessagingConfig()).start().join();
 
-    address2 = Address.from(findAvailablePort(5002));
+    address2 = Address.from(SocketUtil.getNextAddress().getPort());
     netty2 =
         (ManagedMessagingService)
             new NettyMessagingService("test", address2, new MessagingConfig()).start().join();
 
-    addressv11 = Address.from(findAvailablePort(5003));
+    addressv11 = Address.from(SocketUtil.getNextAddress().getPort());
     nettyv11 =
         (ManagedMessagingService)
             new NettyMessagingService("test", addressv11, new MessagingConfig(), ProtocolVersion.V1)
                 .start()
                 .join();
 
-    addressv12 = Address.from(findAvailablePort(5004));
+    addressv12 = Address.from(SocketUtil.getNextAddress().getPort());
     nettyv12 =
         (ManagedMessagingService)
             new NettyMessagingService("test", addressv12, new MessagingConfig(), ProtocolVersion.V1)
                 .start()
                 .join();
 
-    addressv21 = Address.from(findAvailablePort(5005));
+    addressv21 = Address.from(SocketUtil.getNextAddress().getPort());
     nettyv21 =
         (ManagedMessagingService)
             new NettyMessagingService("test", addressv21, new MessagingConfig(), ProtocolVersion.V2)
                 .start()
                 .join();
 
-    addressv22 = Address.from(findAvailablePort(5006));
+    addressv22 = Address.from(SocketUtil.getNextAddress().getPort());
     nettyv22 =
         (ManagedMessagingService)
             new NettyMessagingService("test", addressv22, new MessagingConfig(), ProtocolVersion.V2)
@@ -424,17 +423,5 @@ public class NettyMessagingServiceTest {
     nettyv22.registerHandler(subject, (address, bytes) -> CompletableFuture.completedFuture(bytes));
     response = nettyv12.sendAndReceive(addressv22, subject, payload).get(10, TimeUnit.SECONDS);
     assertArrayEquals(payload, response);
-  }
-
-  private static int findAvailablePort(final int defaultPort) {
-    try {
-      final ServerSocket socket = new ServerSocket(0);
-      socket.setReuseAddress(true);
-      final int port = socket.getLocalPort();
-      socket.close();
-      return port;
-    } catch (final IOException ex) {
-      return defaultPort;
-    }
   }
 }

--- a/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyUnicastServiceTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyUnicastServiceTest.java
@@ -24,8 +24,7 @@ import static org.slf4j.LoggerFactory.getLogger;
 import io.atomix.cluster.messaging.ManagedUnicastService;
 import io.atomix.cluster.messaging.MessagingConfig;
 import io.atomix.utils.net.Address;
-import java.io.IOException;
-import java.net.ServerSocket;
+import io.zeebe.test.util.socket.SocketUtil;
 import net.jodah.concurrentunit.ConcurrentTestCase;
 import org.junit.After;
 import org.junit.Before;
@@ -68,8 +67,8 @@ public class NettyUnicastServiceTest extends ConcurrentTestCase {
 
   @Before
   public void setUp() throws Exception {
-    address1 = Address.from("127.0.0.1", findAvailablePort(5001));
-    address2 = Address.from("127.0.0.1", findAvailablePort(5002));
+    address1 = Address.from("127.0.0.1", SocketUtil.getNextAddress().getPort());
+    address2 = Address.from("127.0.0.1", SocketUtil.getNextAddress().getPort());
 
     service1 = new NettyUnicastService(address1, new MessagingConfig());
     service1.start().join();
@@ -94,18 +93,6 @@ public class NettyUnicastServiceTest extends ConcurrentTestCase {
       } catch (final Exception e) {
         LOGGER.warn("Failed stopping netty2", e);
       }
-    }
-  }
-
-  private static int findAvailablePort(final int defaultPort) {
-    try {
-      final ServerSocket socket = new ServerSocket(0);
-      socket.setReuseAddress(true);
-      final int port = socket.getLocalPort();
-      socket.close();
-      return port;
-    } catch (final IOException ex) {
-      return defaultPort;
     }
   }
 }

--- a/test-util/src/main/java/io/zeebe/test/util/socket/SocketUtil.java
+++ b/test-util/src/main/java/io/zeebe/test/util/socket/SocketUtil.java
@@ -45,7 +45,7 @@ public final class SocketUtil {
 
   private SocketUtil() {}
 
-  public static InetSocketAddress getNextAddress() {
+  public static synchronized InetSocketAddress getNextAddress() {
     return PORT_RANGE.next();
   }
 }

